xapian-applications/omega/handler_gmime.cc

```diff
@@ -0,0 +1,174 @@
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include "myhtmlparse.h"
+#include <string.h>
+
+#include <glib.h>
+#include <gmime/gmime.h>
+
+using namespace std;
+
+const unsigned SIZE = 8192;
+
+static void
+extract_html(const string& text, string& charset, string& dump)
+{
+    MyHtmlParser parser;
+    if (charset.empty())
+	charset = "UTF-8";
+    try {
+	parser.ignore_metarobots();
+	parser.parse_html(text, charset, false);
+	dump += parser.dump;
+    } catch (const string& newcharset) {
+	parser.reset();
+	parser.ignore_metarobots();
+	parser.parse_html(text, newcharset, true);
+	dump += parser.dump;
+    } catch (...) {
+	return;
+    }
+}
+
+size_t decode(unsigned char* text, size_t len, GMimeContentEncoding encoding)
+{
+    unsigned char buffer[SIZE];
+    int state;
+    guint32 save;
+    switch (encoding) {
+	case GMIME_CONTENT_ENCODING_BASE64: {
+	    len = g_mime_encoding_base64_decode_step(text, len, buffer,
+						     &state, &save);
+	    memcpy(text, buffer, len);
+	    break;
+	}
+	case GMIME_CONTENT_ENCODING_UUENCODE: {
+	    len = g_mime_encoding_uudecode_step(text, len, buffer,
+						&state, &save);
+	    memcpy(text, buffer, len);
+	    break;
+	}
+	case GMIME_CONTENT_ENCODING_QUOTEDPRINTABLE: {
+	    len = g_mime_encoding_quoted_decode_step(text, len, buffer,
+						     &state, &save);
+	    memcpy(text, buffer, len);
+	    break;
+	}
+	default:
+	    break;
+    }
+    return len;
+}
+
+bool
+parser_content(GMimeObject* me, string& dump)
+{
+    GMimeContentType* ct = g_mime_object_get_content_type(me);
+    if (GMIME_IS_MULTIPART(me)) {
+	GMimeMultipart* mpart = (GMimeMultipart*)me;
+	string subtype = g_mime_content_type_get_media_subtype(ct);
+	int count = g_mime_multipart_get_count(mpart);
+	if (subtype == "alternative") {
+	    for (int i = 0; i < count; ++i) {
+		GMimeObject* part = g_mime_multipart_get_part(mpart, i);
+		if (parser_content(part, dump))
+		    return true;
+	    }
+	} else {
+	    bool ret = false;
+	    for (int i = 0; i < count; ++i) {
+		GMimeObject* part = g_mime_multipart_get_part(mpart, i);
+		ret |= parser_content(part, dump);
+	    }
+	    return ret;
+	}
+    } else if (GMIME_IS_PART(me)) {
+	GMimePart* part = (GMimePart*)me;
+	GMimeDataWrapper* content = g_mime_part_get_content_object(part);
+	string type = g_mime_content_type_get_media_type(ct);
+	string subtype = g_mime_content_type_get_media_subtype(ct);
+	string charset(g_mime_object_get_headers(me));
+	size_t start = charset.find("charset=");
+	if (start != string::npos) {
+	    start += 9;
+	    charset = charset.substr(start, charset.length() - start - 3);
+	    charset = g_mime_charset_canon_name(charset.c_str());
+	} else {
+	    charset.clear();
+	}
+	if (type == "text") {
+	    char* buffer = new char[SIZE];
+	    GMimeStream* sr = g_mime_data_wrapper_get_stream(content);
+	    size_t presz = 0, len = g_mime_stream_read(sr, buffer, SIZE);
```

Comment:
`presz` seems rather a cryptic name.  Even looking where it is used I'm not really any the wiser.

I'd also avoid declaring multiple variables on the same line if any have non-trivial initialisers - it isn't particularly easy to read.

You also really want to avoid doing this if any of the types are pointers or references - `int* x, y` actually declare `y` as of type `int` - the C++ style of putting the `*` with the type misleads in this case.

Something like `int x = 0, y = 0;` would be OK.